StandaloneMmPkg/Drivers/StandaloneMmIplPei/StandaloneMmIplPei.c

```diff
@@ -8,6 +8,326 @@
+    if (EFI_ERROR (Status)) {
+      break;
+    }
+
+    //
+    // Search PEIM FFS
+    //
+    FileHandle = NULL;
+    Status     = PeiServicesFfsFindNextFile (EFI_FV_FILETYPE_MM_CORE_STANDALONE, VolumeHandle, &FileHandle);
+    if (EFI_ERROR (Status)) {
+      continue;
+    }
+
+    ASSERT (FileHandle != NULL);
+    if (FileHandle != NULL) {
+      *MmCoreFileName = &((EFI_FFS_FILE_HEADER *)FileHandle)->Name;
+    }
+
+    //
+    // Search Section
+    //
+    Status = PeiServicesFfsFindSectionData (EFI_SECTION_PE32, FileHandle, MmCoreImageAddress);
+    if (EFI_ERROR (Status)) {
+      continue;
+    }
+
+    //
+    // Great!
+    //
+    SectionData = (EFI_PE32_SECTION *)((UINT8 *)*MmCoreImageAddress - sizeof (EFI_PE32_SECTION));
+    ASSERT (SectionData->Type == EFI_SECTION_PE32);
+
+    //
+    // This is SMM BFV
+    //
+    Status = PeiServicesFfsGetVolumeInfo (VolumeHandle, &VolumeInfo);
+    if (!EFI_ERROR (Status)) {
+      *MmFvBase = (EFI_PHYSICAL_ADDRESS)(UINTN)VolumeInfo.FvStart;
+      *MmFvSize = VolumeInfo.FvSize;
+      return EFI_SUCCESS;
+    } else {
+      return EFI_NOT_FOUND;
+    }
+  }
+
+  return EFI_NOT_FOUND;
+}
+
+/**
+  Find largest unallocated MMRAM in current MMRAM descriptor block
+
+  @param[in, out] LagestMmramRangeIndex  Lagest mmram range index.
+  @param[in]      CurrentBlock           Current MMRAM descriptor block.
+
+**/
+VOID
+FindLargestSmramRange (
+  IN OUT UINTN                       *LagestMmramRangeIndex,
+  IN EFI_MMRAM_HOB_DESCRIPTOR_BLOCK  *CurrentBlock
+  )
+{
+  UINTN                 Index;
+  UINT64                MaxSize;
+  EFI_MMRAM_DESCRIPTOR  *MmramRanges;
+
+  MmramRanges = CurrentBlock->Descriptor;
+
+  //
+  // Find largest Mmram range.
+  //
+  for (Index = 0, MaxSize = SIZE_256KB - EFI_PAGE_SIZE; Index < CurrentBlock->NumberOfMmReservedRegions; Index++) {
+    //
+    // Skip any MMRAM region that is already allocated, needs testing, or needs ECC initialization
+    //
+    if ((MmramRanges[Index].RegionState & (EFI_ALLOCATED | EFI_NEEDS_TESTING | EFI_NEEDS_ECC_INITIALIZATION)) != 0) {
+      continue;
+    }
+
+    if (MmramRanges[Index].CpuStart >= BASE_1MB) {
+      if ((MmramRanges[Index].CpuStart + MmramRanges[Index].PhysicalSize) <= BASE_4GB) {
+        if (MmramRanges[Index].PhysicalSize >= MaxSize) {
+          DEBUG ((DEBUG_INFO, "Found largest unlocated MMRAM\n"));
+          *LagestMmramRangeIndex = Index;
+          return;
+        }
+      }
+    }
+  }
+
+  DEBUG ((DEBUG_ERROR, "Not found largest unlocated MMRAM\n"));
+  ASSERT (FALSE);
+  CpuDeadLoop ();
+
+  return;
+}
+
+/**
+  Allocate available MMRAM for MM core image.
+
+  @param[in]  Pages                     Page size of MM core image.
+  @param[out] NewBlock                  Pointer of new mmram block HOB.
+
+  @return  EFI_PHYSICAL_ADDRESS         Address for MM core image to be loaded in MMRAM.
+**/
+EFI_PHYSICAL_ADDRESS
+MmIplAllocateSmramPage (

```

Comment:
Better to have the name to be MmIplAllocate**M**mramPage?

Reply:
Updated in new commit